src/tools/list_all_objects/worker.cpp

```diff
@@ -33,6 +35,68 @@ namespace {
 
 constexpr auto token = boost::asio::as_tuple(boost::asio::use_awaitable);
 
+template <s3cpp::tools::list_all_objects::ListObjectsApiVersion api_version>
+[[nodiscard]] s3cpp::meta::crt<boost::asio::awaitable<
+    std::conditional_t<api_version == s3cpp::tools::list_all_objects::ListObjectsApiVersion::V1,
+                       s3cpp::aws::s3::ListObjectsResult, s3cpp::aws::s3::ListObjectsV2Result>>>
+list_one_impl(s3cpp::aws::s3::Client client, std::string bucket, std::optional<std::string> prefix,
+              std::optional<std::string> continuation_token) {
+    const std::string my_prefix = prefix.value_or("<no prefix>");
+    struct Visitor {
+        static std::string operator()(const boost::beast::error_code &error) { return error.what(); }
+        static std::string operator()(const pugi::xml_parse_status &error) {
+            return std::format("pugixml error {}", std::to_underlying(error));
+        }
+    };
+    for (int retries = 0; retries < 5; retries++) {
+        if constexpr (api_version == s3cpp::tools::list_all_objects::ListObjectsApiVersion::V1) {
+            auto res = co_await client.list_objects(
+                {.Bucket = bucket, .Marker = continuation_token, .Delimiter = "/", .Prefix = prefix});
+            if (!res) {
+                const auto errstr = std::visit(Visitor{}, res.error());
+                std::println(std::cerr, "ERROR in prefix {} {} - retrying", my_prefix, errstr);
+                if (std::holds_alternative<boost::beast::error_code>(res.error())) {
+                    continue;
+                }
+            }
+            co_return res.value();
+        } else {
```

Comment:
Unless this is some weird `if constexpr` semantic, do you actually need this else?  After all, in all paths of control flow, you do break the control flow in the if statement above.